lib/ecto/query/planner.ex

```diff
@@ -307,10 +307,52 @@ defmodule Ecto.Query.Planner do
   defp normalize_subquery_select(query, adapter, source?) do
     {schema_or_source, expr, %{select: select} = query} = rewrite_subquery_select_expr(query, source?)
     {expr, _} = prewalk(expr, :select, query, select, 0, adapter)
-    {{:map, types}, _fields, _from} = collect_fields(expr, [], :never, query, select.take, true, %{})
+    {{:map, types}, fields, _from} = collect_fields(expr, [], :never, query, select.take, true, %{})
+    # types must take into account selected_as/2 aliases so that the correct fields are
+    # referenced when the outer query selects the entire subquery
+    types = normalize_subquery_types(types, Enum.reverse(fields), query.select.aliases, [])
     {query, subquery_source(schema_or_source, types)}
   end
 
+  defp normalize_subquery_types(types, _fields, select_aliases, _acc) when select_aliases == %{} do
+    types
+  end
+
+  defp normalize_subquery_types([], [], _aliases, acc) do
+    Enum.reverse(acc)
+  end
+
+  defp normalize_subquery_types([{alias, _} = type | types], [{:selected_as, _, [_, alias]} | fields], select_aliases, acc) do
+    normalize_subquery_types(types, fields, select_aliases, [type | acc])
+  end
+
+  defp normalize_subquery_types([{source_alias, type_value} | types], [field | fields], select_aliases, acc) do
+    if Map.has_key?(select_aliases, source_alias) do
+      raise ArgumentError, """
+      the alias, #{inspect(source_alias)}, provided to `selected_as/2` conflicts with the subquery's
+      `select` statement. Please ensure there is no overlap between the aliases provided to
+      `selected_as/2` and the identifiers used in the `select` statement.
+
+      For instance, the following query is not allowed because the alias `:x2` used in `selected_as/2`
+      is the same as another key in the map being selected:
+
+        s = from(s in Schema, select: %{x: selected_as(s.x, :x2), x2: s.x})
+        from s in subquery(s)
+
+      If the alias provided to `selected_as/2` is changed to any atom other than `:x2`, the above query will work.
+      The alias `:x` would also be allowed because it is the key corresponding to the `selected_as/2` value.
+      """
```

Comment:
Should we just make it so `selected_as` wins, as you originally proposed?

Reply:
Oh what I was trying to say is this:

If you have `select: %{x: selected_as(field, :alias)}` then `:alias` wins over `:x`.

If you have `select: %{x: selected_as(field, :alias), alias: field2}` then `selected_as(field, :alias)` isn't allowed anymore because it conflicts with the other field's alias.

My wording might be unclear. I updated it.

Reply:
Now that I look at this again it isn't perfect. i.e. this shouldn't raise: `select: %{x: selected_as(field, :alias), alias: selected_as(field2, :not_alias)}`.

I think I can change this to do a pass over the over the keys and make them `selected_as` as long as their value isn't `selected_as` something different.

Reply:
Oh, I see.

I think raising on this case is enough:

> If you have select: %{x: selected_as(field, :alias), alias: field2} then selected_as(field, :alias) isn't allowed anymore because it conflicts with the other field's alias.

I think we should also raise on this:

> Now that I look at this again it isn't perfect. i.e. this shouldn't raise: select: %{x: selected_as(field, :alias), alias: selected_as(field2, :not_alias)}.

Because that would be just too confusing anyway.